### ISSUE TITLE
Feature: When the discovery fails it should rather use cached values instead of nothing

### ISSUE THREAD

**Issue by jensbaitingerbosch** [promregator/promregator#116]
## Environment
cf api: 2.120.0 
promregator: 0.5.7

## Summary / Problem Statement
During an incident the CF API was answering very slow. After the cache invalidated, the discovery returned no apps (beside the promregator itselve). 

## Observed Behavior
Since no apps where contained in the discovery, prometheus did not scrape the metrics from the apps and raised alerts that all apps are down. 

## Expected Behavior
Discovery should return the previously cached values so prometheus can continue scraping.

**Comment by eaglerainbow** [comment]
> Discovery should return the previously cached values so prometheus can continue scraping.

I thought on this idea some time back (but did not document it anywhere). 
The problem would be that the cache behavior and all the meanings of the configuration parameters would change in an incompatible way. Thus, this would be a very big step and required a "1.x.y" to be released. 
Moreover, it also has a business implication: So far, consumers of Promregator could rely on after some timeout that the "most recent state" must be in place, or no scraping would happen (indicating an error, which most likely is in place).
Additionally, there is also the problem of new deployments to an application: It can become very tricky to decide whether an old application has "just vanished" because of a bug/issue/timeout of the CF API, or whether the application was deleted intentionally (due the deployment mentioned before) -- you don't want to have old applications still tried to be scraped, which have been deleted long time ago. Thus, we need to get rid of that. CF does not emit events properly well enough for that (e.g. I saw app events being dropped if both the application and the space it belonged to got deleted). 

So in short: yes, nice idea, but bringing it to life without unwanted misbehavior might be very tricky.

**Comment by jensbaitingerbosch** [comment]
The thing is that it does not produce an timeout or http error status code, but it returns a list of target containing only promregator itself: 

```
[
  {
    "targets": [
      "40bb18c5-1c99-4c63-6180-3f97:8080"
    ],
    "labels": {
      "__meta_promregator_target_path": "/promregatorMetrics",
      "__meta_promregator_target_orgName": null,
      "__meta_promregator_target_spaceName": null,
      "__meta_promregator_target_applicationName": null,
      "__meta_promregator_target_applicationId": null,
      "__meta_promregator_target_instanceNumber": null,
      "__meta_promregator_target_instanceId": null,
      "__metrics_path__": "/promregatorMetrics"
    }
  }
]
```

This is taken by my discovery scraping job and presented to prometheus which then stops scraping the apps. 

To Implement this I whould recommend some kind of 2 state cache: 
(pseudocode)
```
if (!cache.timedout) {
  return cache.content
} else {
  try {
    return cfapi.query()
  } catch (TimoutException | ServerErrorException ex) {
    return cache.content
  }
}

**Comment by eaglerainbow** [comment]
With 0f3089817c7ff51edce09db3c459fa39a2db024f (branch `guavaCache`) I have been playing around with guava as central cache replacement. However, it turned out that the enforced synchronous loading of Guava did not work out with the asynchronous loading behaviour of Monos. I did not want to have a cache with value types `Mono<?>`, as this would have deactivated 90% of the caching logic (and we wouldn't have got what we wanted to have - then the values would have been refreshed even though the result of the Mono wasn't computed yet - and hence could error). 

Thanks to a SO article, I came across Caffeine, which is very similar to Guava (from the API), but has an `AsyncLoadingCache`, which works with `CompletableFuture`s. These are 1:1 castable to Monos. Caffeine also has Prometheus' simple_client support out of the box. At the current state (9da6b18a5c3f2ffc7017cbc0a64524d59171d01b), the metrics are provided via `InternalMetrics` already.

Still to do:
* Get more experience how good the cache is
* Unit Tests are missing completely.

**Comment by eaglerainbow** [comment]
Observation: cache entries subject to `refreshAfterWrite` are not auto-refreshed, but only after next access. Apparently, loading then is triggered synchronously (e.g. the cache requestor is blocked until the future of "async load" could be resolved).

**Comment by eaglerainbow** [comment]
Observation: to what @jensbaitingerbosch wants to have, `expireAfterAccess` would have to be set high, as with low values (or values just short above the refresh timing), still the records would be evicted by the cache automatically.
Internal warning: We cannot drop `expireAfterAccess` altogether! Cache entries for apps, which are removed by the CF API/CLI, must get evicted at some point in time! The only alternative for this would be other eviction policies (such as max record size), but this would be again very hard to configure properly.

**Comment by eaglerainbow** [comment]
> Get more experience how good the cache is

Some tests show that the cache really is good: They can perform real asynchronous, while still handing out old instances, which got resolved properly. 
Also combining it with reactor is good, as `Mono`s can be converted to `CompletableFuture`s back and forth. 
A little tricky is the topic with `subscribeOn`: Caffeine demand to have resolutions of Futures to be executed on a special `Executor`. However, the cf_java library has its own thoughts about where to execute what.

Eventually, with d22a49d8bbd220b9dccab15790cc711df25c3a4a I was able to get a call to `/discovery` on a refresh-request-in-queue situation responded completely from the state of the (stale, but data-loading-pending) cache. That's rather impressive.

I am also rather pleased about the fact that I got it implemented in a rather compatible way: We just have to replace the entire `CFAccessorCache` - but that can be done with a configuration switch. 
Hence, we have to carefully consider what we write into the release notes + the documentation.

**Comment by ben-manes** [comment]
> A little tricky is the topic with subscribeOn: Caffeine demand to have resolutions of Futures to be executed on a special Executor. However, the cf_java library has its own thoughts about where to execute what.

Can you explain what you mean? We pass all asynchronous logic to the configured `Caffeine.executor` and forward that along to for optional use by async computations. If it is used, or not, is your choice. By default we use `ForkJoinPool.commonPool()`, but you aren't restricted to that (e.g. `Runnable::run` is handy for unit tests to disable any async behavior).

You might also be interested in `Caffeine.scheduler`, perhaps using `Schedulers.systemScheduler()`, to avoid your `guavaCacheCleanup` method.

**Comment by eaglerainbow** [comment]
@ben-manes 

> A little tricky is the topic with subscribeOn: Caffeine demand to have resolutions of Futures to be executed on a special Executor

Ah, that's perhaps abbreviated badly by me.  Let me try to explain it better...

> We pass all asynchronous logic to the configured Caffeine.executor and forward that along to for optional use by async computations. If it is used, or not, is your choice.

That's perhaps the point here: In my special case that I have, I *must* use it. Otherwise the main thread might get blocked, and I don't get the asynchronous behavior of the cache properly (there is still some reactive stuff going on there, which has "more priority").
That the asynchronous loading is handed over to the `ForkJoinPool.commonPool()` is fine for me - it just must not be my main thread 😁 

> You might also be interested in Caffeine.scheduler, perhaps using Schedulers.systemScheduler(), to avoid your guavaCacheCleanup method.

Thanks for the pointer - sounds good and I'll check if that will work in my case.

BTW: Thanks for jumping in here, even though I did not "mention" your repo directly! 
It's great to get early feedback even in the state of yet experimenting with your lib!

**Comment by eaglerainbow** [comment]
Ah, snap: `Caffeine.scheduler` is 2.8+ only 😞 
Due to other dependencies, I am currently locked into 2.6.2

**Comment by ben-manes** [comment]
Oh, too bad. You probably want 2.7 for bug fixes, though, and we have good backwards compatibility.

You might also consider using Awaitility for unit tests to avoid explicit Thread.sleep calls. It’s pretty handy for asynchronous testing.

**Comment by eaglerainbow** [comment]
> Oh, too bad. You probably want 2.7 for bug fixes, though, and we have good backwards compatibility.

Thanks again for the pointer! I'll check how hard the dependency lock is. So far, I didn't encounter any bug that would make me cry up, though... Your stuff is really good - it's more that I apparently have to still learn a little more, how to get the right out of it.

> You might also consider using Awaitility for unit tests to avoid explicit Thread.sleep calls. It’s pretty handy for asynchronous testing.

Yep, saw it. 385aa84f24a06e23021c8569e1b44e6497baee41 already makes use of `FakeTicker`

**Comment by eaglerainbow** [comment]
Arg, too bad: Caffeine in the standard variant also keeps broken/`Mono.error` instances in the cache.

**Comment by ben-manes** [comment]
It should remove futures that complete exceptionally or with a null result. However that is via a whenComplete callback.

**Comment by eaglerainbow** [comment]
🤔 
Then I don't understand why the adoption of the timeout unit tests (see 4f514d852882c882217ddbeed0a53139313d86d5) does not work (yet) - but perhaps that my local time and my brain both me "go to bed" 😁 
I'll have to look at this case tomorrow, when there is more power on my brain again.
Thanks for all your help! Really appreciated!

**Comment by ben-manes** [comment]
Perhaps a race of the test thread reading before the handler can remove the entry?

goodnight!

**Comment by eaglerainbow** [comment]
> Perhaps a race of the test thread reading before the handler can remove the entry?

Something like this.
I have more the impression that the second request was calling too fast shortly after the first still was pending. Introducing a short delay (915a7ecdd0f2622f4692984f07573bc232caa354) apparently did the trick.

**Comment by ben-manes** [comment]
This is where Awaitility is nice, e.g.
```java
await().until(cache.getIfPresent(key), is(nullValue());
```

**Comment by eaglerainbow** [comment]
> `scheduler()`: Thanks for the pointer - sounds good and I'll check if that will work in my case.

I managed to bump the Caffeine version with less of a problem than I had thought. However, Promregator is still on JDK8, so I couldn't really use `Scheduler.systemScheduler()`. Instead, I created an own pool for that (for Promregator that's anyway ok).

**Comment by eaglerainbow** [comment]
> This is where Awaitility is nice, e.g.
> `await().until(() -> cache.getIfPresent(key), is(nullValue());`

mmmh.. I tend to agree; unfortunately, the unit test does not have access to the cache reference directly. Permitting access would mean to break up the interface of the class under test 🤔

**Comment by eaglerainbow** [comment]
Just for the record: We still need documentation on `cf.cache.type` for a final release

**Comment by eaglerainbow** [comment]
I just released [v0.7.0-rc1](https://github.com/promregator/promregator/releases/tag/v0.7.0-rc1) with Caffeine support.
@warlockza @jensbaitingerbosch Please feel free to test this out (in a non-prod environment). Feedback is highly appreciated!

**Comment by eaglerainbow** [comment]
Just for the record: 0.7.1-rc1 is running on a test machine on caffeine cache for 24hrs+ now. So far, I could not make out any caching issue.

**Comment by eaglerainbow** [comment]
Just for the record: 0.7.1-rc1 is running fine on a test machine with caffeine cache enabled for more than 4 days now.

**Comment by warlockza** [comment]
@eaglerainbow how is the test coverage on the new code progressing?
https://sonarcloud.io/component_measures?id=promregator&metric=new_coverage&view=list . Sonar is bleating :)

**Comment by eaglerainbow** [comment]
> how is the test coverage on the new code progressing?

Coverage of the Caffeine cache is very good:
![grafik](https://user-images.githubusercontent.com/12811079/70347312-91c8bb80-1860-11ea-84d9-bef403186beb.png)

So, in terms of LoC coverage: it's 97%. Hence, it's not the Caffeine cache which makes Sonar cry. 😁 

The real culprit is https://sonarcloud.io/component_measures?id=promregator&metric=new_coverage&selected=promregator%3Asrc%2Fmain%2Fjava%2Forg%2Fcloudfoundry%2Fpromregator%2Fcfaccessor%2FReactiveCFAccessorImpl.java&view=list

which you touched. The class is the low-level communication handler towards Cloud Foundry. It's very hard to unit-test it, as it required to mock all the complex calls to the Cloud Foundry Cloud Controller (CFCC). I did not come around with a magnificent idea for it yet.

**Comment by eaglerainbow** [comment]
Given that I have been running a Promregator instance on Caffeine for more than 4 weeks now and it did not cause any cache-related incident (though there were plenty of chances), I'll release 0.7.1 in a moment.

**Comment by eaglerainbow** [comment]
Here we go: https://github.com/promregator/promregator/releases/tag/v0.7.1
